target/riscv/pmp.c

```diff
@@ -245,7 +245,7 @@ bool pmp_hart_has_privs(CPURISCVState *env, target_ulong addr,
          from low to high */
     for (i = 0; i < MAX_RISCV_PMPS; i++) {
         s = pmp_is_in_range(env, i, addr);
-        e = pmp_is_in_range(env, i, addr + size);
+        e = pmp_is_in_range(env, i, addr + size - 1);
```

Comment:
This doesn't look right. This fixes the less then end address but breaks the greater then or equal to start address.

Shouldn't the diff be this instead?
diff --git a/target/riscv/pmp.c b/target/riscv/pmp.c
index f432f3b759..9eae74715b 100644
--- a/target/riscv/pmp.c
+++ b/target/riscv/pmp.c
@@ -210,7 +210,7 @@ static int pmp_is_in_range(CPURISCVState *env, int pmp_index, target_ulong addr)
     int result = 0;
 
     if ((addr >= env->pmp_state.addr[pmp_index].sa)
-        && (addr <= env->pmp_state.addr[pmp_index].ea)) {
+        && (addr < env->pmp_state.addr[pmp_index].ea)) {
         result = 1;
     } else {
         result = 0;

Reply:
Thank you for the review!

However, that will make even more legitimate accesses fail, and also will make it return 0 for some in-PMP-range accesses.
Because `env->pmp_state.addr[pmp_index].ea` is set to the last address, not a boundary address,
even a `pmp_is_in_range` call with the last address itself will always fail.

Let me give you an example, assuming your diff has been applied.
If PMP range is from 0x0010 to 0x0020, `sa` will set to 0x0010 and `ea` will set to 0x001f
Accessing 1 byte at 0x001f will call
`pmp_is_in_range(0x001f)` and 
`pmp_is_in_range(0x0020)`
Since both 0x001f and 0x0020 are not less than `ea` (=0x001f), both call will return 0,
meaning that the address is not in the PMP range, which is not true.

I think `pmp_is_in_range` is doing great, but the issue here is that we are using a wrong address.
Can you explain more about why does the change break the greater then or equal to start address rule?

Reply:
Looking at page 47 on the privileged spec I see " the entry matches any address a such that pmpaddri-1 <= a < pmpaddri" which is why I think the pmp_is_in_range() should be changed to less then.

Although I do see your point. It looks like ea already has one subtracted from it when it is set so you are right about it being the boundary.

Reply:
Should that 1 not be subtracted when setting ea instead? Then we can have the same comparison that the spec has

Reply:
I agree. Not subtracting 1 from `ea` should also work.
But note that the spec saying pmpaddri-1 <= a < pmpaddri will be satisfied with both implementations. 
pmpaddri is a riscv register value whereas `ea` is a qemu's variable to configure a pmp range end address.

Reply:
I found that not substracting 1 from `ea` will complicate the original code
especially when we want to include the entire memory inside a range.
Currently we're assigning -1 (all ones) no matter how large the actual physical memory is,
but if we add 1, then we should know the actual memory size.
Better way around is just pass the end address to `pmp_is_in_range`, 
which is `base + size - 1`

Reply:
Ok, in which case all of this looks good to me.

You can add my:
Reviewed-by: Alistair Francis <alistair.francis@wdc.com>
Just copy it below your SOB line. It would be good if you could still split the two patches

Reply:
Hi @alistair23,
Should I just PR this again to this repo (riscv-qemu), or report it to upstream QEMU?

Reply:
I think you should do both. It makes sense to have it merged here (then it's in and people can use it). It also will eventually need to go upstream so why not do it now :)

QEMU is in freeze though, so it won't be merged straight away.